zebra-state/src/service/check/anchors.rs

```diff
@@ -214,119 +411,68 @@ pub(crate) fn sprout_anchors_refer_to_treestates(
-
-            tracing::debug!(
-                ?joinsplit.anchor,
-                ?joinsplit_index_in_tx,
-                ?tx_index_in_block,
-                ?height,
-                "validated sprout anchor",
-            );
+    block
+        .transactions
+        .par_iter()
+        .enumerate()
+        .try_for_each(|(tx_index_in_block, transaction)| {
+            sprout_anchors_refer_to_treestates(
+                &sprout_final_treestates,
+                transaction,
+                transaction_hashes[tx_index_in_block],
+                Some(tx_index_in_block),
+                Some(height),
+            )?;
+
+            Ok(())
+        })
+}
 
-            // The last interstitial treestate in a transaction can never be used,
-            // so we avoid generating it.
-            if joinsplit_index_in_tx == joinsplit_count - 1 {
-                continue;
-            }
+/// Accepts a [`ZebraDb`], an optional [`Option<Arc<Chain>>`](Chain), and an [`UnminedTx`].
+///
+/// Checks the final Sprout, Sapling and Orchard anchors specified in the [`UnminedTx`].
+///
+/// This method checks for anchors computed from the final treestate of each block in
+/// the `parent_chain` or `finalized_state`.
+#[tracing::instrument(skip_all)]
+pub(crate) fn tx_anchors_refer_to_final_treestates(
+    finalized_state: &ZebraDb,
+    parent_chain: Option<&Arc<Chain>>,
+    unmined_tx: &UnminedTx,
+) -> Result<(), ValidateContextError> {
+    sapling_orchard_anchors_refer_to_final_treestates(
+        finalized_state,
+        parent_chain,
+        &unmined_tx.transaction,
+        unmined_tx.id.mined_id(),
+        None,
+        None,
+    )?;
 
-            let input_tree_inner = Arc::make_mut(&mut input_tree);
+    let mut sprout_final_treestates = HashMap::new();
 
-            // Add new anchors to the interstitial note commitment tree.
-            for cm in joinsplit.commitments {
-                input_tree_inner
-                    .append(cm)
-                    .expect("note commitment should be appendable to the tree");
-            }
+    fetch_sprout_final_treestates(
+        &mut sprout_final_treestates,
+        finalized_state,
+        parent_chain,
+        &unmined_tx.transaction,
+        None,
+        None,
+    );
 
-            interstitial_trees.insert(input_tree.root(), input_tree);
+    tracing::trace!(
+        sprout_final_treestate_count = ?sprout_final_treestates.len(),
+        ?sprout_final_treestates,
+        "received sprout final treestate anchors",
+    );
 
-            tracing::debug!(
-                ?joinsplit.anchor,
-                ?joinsplit_index_in_tx,
-                ?tx_index_in_block,
-                ?height,
-                "observed sprout interstitial anchor",
-            );
-        }
-    }
+    sprout_anchors_refer_to_treestates(
+        &sprout_final_treestates,
+        &unmined_tx.transaction,
+        unmined_tx.id.mined_id(),
+        None,
+        None,
+    )?;
```

Comment:
Like the block note commitment tree updates, I think we might want to run this verification on a CPU-bound thread pool.

It would go something like this, but maybe it will need some clones.
```suggestion
    // If there are no sprout transactions in the block, avoid running a rayon scope for checks that will never fail.
    if sprout_final_treestates.is_empty() {
        return Ok(());
    }
    
    // This check is expensive, because it updates a note commitment tree for each sprout JoinSplit.
    // Since we could be processing attacker-controlled mempool transactions, we need to run each one
    // in its own thread, separately from tokio's blocking I/O threads. And if we are under heavy load,
    // we want verification to finish in order, so that later transactions can't delay earlier ones.
    rayon::scope_fifo(move |s| {
        s.spawn_fifo(move |s| {
            sprout_anchors_refer_to_treestates(
                &sprout_final_treestates,
                &unmined_tx.transaction,
                unmined_tx.id.mined_id(),
                None,
                None,
            )
        })
    })?;
```

Reply:
I couldn't find your comment about it, but I think you're right: the function we need to use is `unmined_tx.transaction.has_sprout_joinsplit_data()`, not the looked up anchors.

Reply:
I figured this was the idea right after commenting and realizing the intent was to skip the rayon scope.

Do you think this state request would often be the longest running check among the async checks? Why is it a DoS risk?

Reply:
> Do you think this state request would often be the longest running check among the async checks?

During normal operation, no, because Sprout transactions are rare.

During an attack that uses lots of Sprout JoinSplits, yes. (But we haven't seen one of those yet.)

> Might it be worth running the duplicate nullifier checks in parallel too?

No, retrieving 32 bytes from disk is usually cheap enough to be run serially, because there are usually 0-2 nullifiers revealed in a transaction, and RocksDB stores a hash index in-memory for nullifiers.

And remember that this request is already running in parallel with the proofs and the other async checks run by the transaction verifier.

> Why is it a DoS risk?

Each Sprout interstitial anchor requires updating the Sprout note commitment tree, which is expensive, because it does `O(n*log(n))` cryptographic operations per `JoinSplit`.

This is a denial of service risk because the mempool verifies remotely-controlled transaction data, without any other proof of work.

If we put that much CPU-bound work into the tokio thread pool, it could slow down network and disk requests.